File: lib/face_native/ios/Classes/data/ImagesVectorDB.swift
import ObjectBox
import Foundation

class ImagesVectorDB {
    // MARK: - Singleton
    static let shared = ImagesVectorDB()
    
    private init() {}
    
private func getImagesBox() throws -> Box<FaceImageRecord> {
        do {
            return try ObjectBoxStore.store.box(for: FaceImageRecord.self)
        } catch {
            print("❌ Error getting ImagesBox:")
                   print("➡️ Type: \(type(of: error))")
                   print("➡️ Description: \(error.localizedDescription)")
                   print("➡️ Full error: \(error)")
                   throw error
        }
    }
    
    func addFaceImageRecord(record: FaceImageRecord) throws -> Id {
        let box = try getImagesBox()    
        let entityId = try box.put(record)
        return entityId.value
    }
    
    func addAllRecords(records: [FaceImageRecord]) throws {
        let box = try getImagesBox()
        _ = try box.put(records)
    }
    
    // Get all records where faceEmbedding is not empty
    func getAllRecords() throws -> [FaceImageRecord] {
        let box = try getImagesBox()
        return try box.all()
    }
    

    func getNearestEmbeddingPersonName(
        embedding: [Float],
        thresholdScore: Float = 0.1
    ) throws -> FaceImageRecord? {
        let box = try getImagesBox()
  
        let query = try box.query {
            FaceImageRecord.faceEmbedding.nearestNeighbors(
                queryVector: embedding,
                maxCount: 10
            )
        }
        .build()
        let results = try query.findWithScores()
        print("📊 Nearest neighbor search returned \(results.count) results")
        for (index, result) in results.prefix(5).enumerated() {
            print("  \(index + 1). Score: \(String(format: "%.4f", result.score)) - \(result.object.personName) (EmpID: \(result.object.empId))")
        }
        
        let bestMatch = results
            .filter { $0.score <= Double(thresholdScore) }
            .min(by: { $0.score < $1.score })
        
        if let match = bestMatch {
            print("✅ Best match: \(match.object.personName) with score \(String(format: "%.4f", match.score))")
        } else {
            print("❌ No match found below threshold \(thresholdScore)")
        }
        
        return bestMatch?.object
    }
    
    func removeFaceRecordsWithEmpId(empId: Int64) throws {
       do {
         let box = try getImagesBox()
        let query = try box.query {
            FaceImageRecord.empId == (empId)
        }
            .build()
        let ids = try query.findIds()
        for id in ids {
            try box.remove(id)
        }
       }
       catch {
        print("Error: removeFaceRecordsWithEmpId failed !!")
        print("Error: \(error.localizedDescription)")
        throw error
       }
    }
    
    func getFaceImageRecordByListEmpId(empIdList: [Int64]) throws -> [FaceImageRecord] {
        let box = try getImagesBox()
        let allRecords = try box.all()
        return allRecords.filter { empIdList.contains($0.empId) }
    }
    
    func getImageIdsByEmpId(empId: Int64) throws -> [Int64] {
        let box = try getImagesBox()
        let allRecords = try box.all()
        return allRecords.filter { $0.empId == empId }.map { Int64($0.id.value) }
    }
    
    func removeFaceRecordsByIds(imageIds: [Int64]) throws {
        let box = try getImagesBox()
        let ids = imageIds.map { Id($0) }
        try box.remove(ids)
    }
    
    // Clear all records
    func clearAllRecords() throws {
        let box = try getImagesBox()
        try box.removeAll()
    }
    
    // Get the count of all records
    func getCount() throws -> Int64 {
        let box = try getImagesBox()
        return Int64(try box.count())
    }
    
}

File: lib/face_native/android/src/main/kotlin/com/example/face_native/data/ImagesVectorDB.kt
package com.example.face_native.data

import android.util.Log
import org.koin.core.annotation.Single
import com.example.face_native.data.FaceImageRecord

@Single
class ImagesVectorDB {
    private val imagesBox get() = ObjectBoxStore.store.boxFor(FaceImageRecord::class.java)

    fun addFaceImageRecord(record: FaceImageRecord): Long {
        val id= imagesBox.put(record)
        return id;
    }
    
    // get all records where faceEmbedding is not null
    fun getAllRecords(): List<FaceImageRecord> {
        return imagesBox
            .query(FaceImageRecord_.faceEmbedding.notNull())
            .build()
            .find()
    }

    fun getNearestEmbeddingPersonName(
        embedding: FloatArray,
        thresholdScore: Float = 0.1f,
    ): FaceImageRecord? {
        // Get nearest neighbors and filter by threshold
        // Lower score = closer distance = better match
        val results = imagesBox
            .query(FaceImageRecord_.faceEmbedding.nearestNeighbors(embedding, 150))
            .build()
            .findWithScores()
            .onEach { println("Score: ${it.score}") }
            .filter { it.score <= thresholdScore } // Keep only good matches
        
        // Return the record with the LOWEST score (closest match)
        val bestMatch = results.minByOrNull { it.score }
        if (bestMatch != null) {
            Log.d("ImagesVectorDB", "Best match score: ${bestMatch.score}")
        } else {
            Log.d("ImagesVectorDB", "No match found below threshold: $thresholdScore")
        }
        return bestMatch?.get()
    }

    fun removeFaceRecordsWithEmpId(empId: Long) {
        imagesBox.removeByIds(
            imagesBox
                .query(FaceImageRecord_.empId.equal(empId))
                .build()
                .findIds()
                .toList(),
        )
    }
    fun getFaceImageRecordByListEmpId(empIdList: List<Long>): List<FaceImageRecord> {
        return imagesBox
            .query(FaceImageRecord_.empId.oneOf(empIdList.toLongArray()))
            .build()
            .find()
    }
    fun getImageIdsByEmpId(empId: Long): List<Long> {
        return imagesBox
            .query(FaceImageRecord_.empId.equal(empId))
            .build()
            .findIds()
            .toList()
    }
    fun removeFaceRecordsByIds(imageIds: List<Long>) {
        imagesBox.removeByIds(imageIds)
    }
    // clear all records
    fun clearAllRecords() {
        imagesBox.removeAll()
    }

    // get the count of all records
    fun getCount(): Long {
        return imagesBox.count()
    }
    // get the size (in bytes) of the database
    fun getDatabaseSizeInBytes(): Long {
        return try {
            val boxStore = imagesBox.store
            boxStore.dbSize
        } catch (e: Exception) {
            0L // Return 0 if unable to get size
        }
    }
}
